core/src/avm2/verify.rs

```diff
@@ -561,279 +670,607 @@ fn optimize<'gc>(
+                {
+                    *op = Op::Nop;
+                }
+                stack.push_int();
+            }
+            Op::CoerceU => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Uint)) {
+                    *op = Op::Nop;
+                }
+                stack.push_uint();
+            }
+            Op::CoerceA => {
+                stack.pop();
+                stack.push_any();
+            }
+            Op::CoerceS => {
+                let stack_value = stack.pop();
+                if matches!(stack_value, Some(ValueType::Null)) {
+                    *op = Op::Nop;
+                }
+                stack.push_class_object(activation.avm2().classes().string);
+            }
+            Op::Equals | Op::LessEquals | Op::LessThan | Op::GreaterThan | Op::GreaterEquals => {
+                stack.pop();
+                stack.pop();
+                stack.push_boolean();
+            }
+            Op::Not => {
+                stack.pop();
+                stack.push_boolean();
+            }
+            Op::PushTrue | Op::PushFalse => {
+                stack.push_boolean();
+            }
+            Op::PushNull => {
+                stack.push_null();
+            }
+            Op::PushUndefined => {
+                stack.push_any();
+            }
+            Op::PushNaN => {
+                stack.push_number();
+            }
+            Op::PushByte { value } => {
+                if *value >= 0 {
+                    stack.push_uint();
+                } else {
+                    stack.push_int();
+                }
+            }
+            Op::PushShort { value } => {
+                if *value >= 0 {
+                    stack.push_uint();
+                } else {
+                    stack.push_int();
+                }
+            }
+            Op::PushInt { value } => {
+                if *value < -(1 << 28) || *value >= (1 << 28) {
+                    stack.push_number();
+                } else if *value >= 0 {
+                    stack.push_uint();
+                } else {
+                    stack.push_int();
+                }
+            }
+            Op::DecrementI => {
+                // This doesn't give any Number-int guarantees
+                stack.pop();
+                stack.push_any();
+            }
+            Op::IncrementI => {
+                // This doesn't give any Number-int guarantees
+                stack.pop();
+                stack.push_any();
+            }
+            Op::DecLocalI { index } => {
+                if (*index as usize) < local_types.len() {
+                    // This doesn't give any Number-int guarantees
+                    local_types.set_any(*index as usize);
+                }
+            }
+            Op::IncLocalI { index } => {
+                if (*index as usize) < local_types.len() {
+                    // This doesn't give any Number-int guarantees
```

Comment:
same as above: the fact that runtime value can be an int or float, doesn't mean that we can't still set it as being int-like.